### PR TITLE
[Merged by Bors] - feat: `S ^ n = S` where `S` is a submonoid

### PR DESCRIPTION
**Pull request by YaelDillies**
From GrowthInGroups

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Subgroup/Pointwise.lean [hunk 1]
```diff
@@ -50,16 +50,30 @@ lemma op_smul_coe_set [Group G] [SetLike S G] [SubgroupClass S G] {s : S} {a : G
...
+namespace Set
+
+open Subgroup
+
+@[to_additive (attr := simp)]
+lemma mul_subgroupClosure (hs : s.Nonempty) : s * closure s = closure s := by
+  rw [← smul_eq_mul, ← Set.iUnion_smul_set]
+  have h a (ha : a ∈ s) : a • (closure s : Set G) = closure s :=
+    smul_coe_set <| subset_closure ha
+  simp (config := {contextual := true}) [h, hs]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Subgroup/Pointwise.lean:68]:
```suggestion
  simp +contextual [h, hs]
```

File: Mathlib/Algebra/Group/Submonoid/Pointwise.lean [hunk 2]
```diff
@@ -53,9 +53,20 @@ way to `Module`).
...
 variable [Monoid M] [AddMonoid A]
 
+@[to_additive (attr := simp, norm_cast)]
+lemma coe_mul_coe [SetLike S M] [SubmonoidClass S M] (H : S) : H * H = (H : Set M) := by
+  aesop (add simp mem_mul)
+
+set_option linter.unusedVariables false in
+@[to_additive (attr := simp)]
+lemma coe_set_pow [SetLike S M] [SubmonoidClass S M] :
+    ∀ {n} (hn : n ≠ 0) (H : S), (H ^ n : Set M) = H
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Submonoid/Pointwise.lean:66]: There doesn't seem to be a need for this:
```suggestion
@[to_additive (attr := simp)]
lemma coe_set_pow [SetLike S M] [SubmonoidClass S M] :
    ∀ {n} (_ : n ≠ 0) (H : S), (H ^ n : Set M) = H
```

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Algebra/Group/Subgroup/Pointwise.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Group/Subgroup/Pointwise.lean:68]: Well spotted!

File: Mathlib/Algebra/Group/Subgroup/Pointwise.lean [hunk 3]
```diff
@@ -50,16 +50,30 @@ lemma op_smul_coe_set [Group G] [SetLike S G] [SubgroupClass S G] {s : S} {a : G
...
+
+@[to_additive (attr := simp)]
+lemma mul_subgroupClosure (hs : s.Nonempty) : s * closure s = closure s := by
+  rw [← smul_eq_mul, ← Set.iUnion_smul_set]
+  have h a (ha : a ∈ s) : a • (closure s : Set G) = closure s :=
+    smul_coe_set <| subset_closure ha
+  simp +contextual [h, hs]
+
+@[to_additive (attr := simp)]
+lemma mul_subgroupClosure_pow (hs : s.Nonempty) : ∀ n, s ^ n * closure s = closure s
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Group/Subgroup/Pointwise.lean]: Well spotted. Also made me notice I accidentally lost half of my lemmas when PRing...

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/Algebra/Group/Subgroup/Pointwise.lean [hunk 4]
```diff
@@ -50,16 +50,43 @@ lemma op_smul_coe_set [Group G] [SetLike S G] [SubgroupClass S G] {s : S} {a : G
...
+  rw [← smul_eq_mul, ← Set.iUnion_smul_set]
+  have h a (ha : a ∈ s) : a • (closure s : Set G) = closure s :=
+    smul_coe_set <| subset_closure ha
+  simp +contextual [h, hs]
+
+open scoped RightActions in
+@[to_additive (attr := simp)]
+lemma subgroupClosure_mul (hs : s.Nonempty) : closure s * s = closure s := by
+  rw [← Set.iUnion_op_smul_set]
+  have h a (ha : a ∈ s) :  (closure s : Set G) <• a = closure s :=
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Subgroup/Pointwise.lean:74]:
```suggestion
  have h a (ha : a ∈ s) : (closure s : Set G) <• a = closure s :=
```

Comment (YaelDillies) [reply Mathlib/Algebra/Group/Subgroup/Pointwise.lean:74]: Whoops, sorry, will fix in the next PR

Comment (YaelDillies) [reply Mathlib/Algebra/Group/Subgroup/Pointwise.lean:74]: #19577

**Pull request closed without merging**